File: Applied Course/4.Problem Solving/4.Problems on Stacks and Queues/5.Stock Span Problem With using stack.py
import sys
class node: 

    def __init__(self, info): 
        self.info = info  
        self.next = None 


class Stack: 

    def __init__(self): 
        self.top = None
        
    
    def isEmpty(self):
        if self.top is None:
            return True
        return False
    
    def push(self,data):
        self.temp=node(data)
        if self.temp is None:
            print("Stack overflow")
            return
        self.temp.next=self.top
        self.top=self.temp
    
    def pop(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        d=self.top.info
        self.top=self.top.next    
        return d
    
    def peek(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        d=self.top.info
        return d
    def display(self):
        if self.isEmpty():
            print("Stack Underflow")
            sys.exit(0)
        self.p=self.top
        while self.p is not None:
            print(self.p.info)
            self.p=self.p.next
def calculateSpan(price,n,stock):
    s=Stack()
    s.push(0)
    stock[0]=1
    
    for i in range(1,n):
        
        while s.isEmpty() is not True and price[s.peek()]<=price[i]:
            ele=s.pop()
        
        if s.isEmpty():
            stock[i]=i+1
        else:
            stock[i]=i-s.peek()
        
        s.push(i)
    return stock


if __name__=='__main__':
    price=[10, 4, 5, 90, 120, 80]
    n=len(price)
    stock=[0]*n
    stock=calculateSpan(price,n,stock)
    for i in stock:
        print("{} ".format(i)),

File: Applied Course/4.Problem Solving/4.Problems on Stacks and Queues/5.Stock Span Problem With using stack.c
#include <stdio.h>
#include <stdlib.h>

// initialize top =-1
int top = -1;

// function to check the stack is empty or not
int isStackEmpty()
{
	return (top == -1);
}

// function to push an element into the stack

void push(int stack[], int data, int size)
{
	if(top == size)
		printf("stack Full");
	else
		stack[++top] = data;
}

// FUnction to pop an element from the stack
int pop(int *stack)
{
	int ele ;
	if(isStackEmpty())
	{
		printf("stack Empty\n");
		return -1;
	}
	ele = stack[top];
	top--;
	return ele;
}

void calculateSpan(int *price, int n, int *stock)
{
	int stack[n], ele;
	push(stack, 0, n);
	
	//span value of first day is always 1
	stock[0] = 1;
	
	//Calculate span value of remaining days by checking previous days
	for(int i=1; i<n; i++)
	{
		//Traverse array left while the next element  on left is smaller than price[i] 
		while(!isStackEmpty() && price[stack[top]] <= price[i])
			ele = pop(stack);

		stock[i] = (isStackEmpty() ? (i+1) : (i- stack[top]));
		push(stack, i, n);
	}
}

// Function to print span on each day
void printSpan(int stock[], int n)
{
	for(int i=0; i<n; i++)
		printf("%d  ", stock[i]);
}

int main()
{
	int price[] = {10, 4, 5, 90, 120, 80};
	int stock[6];
	calculateSpan(price, 6, stock);
	printSpan(stock, 6);
	return 0;
}